### PR TITLE
Loosen type condition in linear, convolution_2d and deconvolution_2d

### LINKED ISSUE #519: Convolution2D does not support 64 bit
When I change dtype=numpy.float32 to dtype=numpy.float64 in the constructor of Convolution2D, and then run nosetests I get tons of errors. I think 64 bit is simply not supported, so we should raise a NotImplementedError.

### PR DESCRIPTION
**Pull request by okuta**
These fcuntions accept an array of numpy.float16, numpy.float32 and numpy.float64.
See also #555.
Fix #519.

### CODE REVIEW

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -218,8 +227,12 @@ def backward_gpu(self, inputs, grad_outputs):
             W_mat = W.reshape(out_c, -1)
             gcol = cuda.cupy.empty_like(self.col)
             gcol_mats = gcol.reshape(n, c * kh * kw, out_h * out_w)
+
             for i in moves.range(n):
-                cuda.cupy.dot(W_mat.T, gy_mats[i], gcol_mats[i])
+                gcol_mats[i] = cuda.cupy.dot(W_mat.T, gy_mats[i])
+
+            gcol = cuda.cupy.tensordot(W, gy, (0, 1)).astype(x.dtype)
+            gcol = cuda.cupy.rollaxis(gcol, 3)
```
Comment (delta2323) [inline chainer/functions/connection/convolution_2d.py]: Could you explain why we need additional mathematical operations, although changes so far are related to dtypes?

File: chainer/functions/connection/convolution_2d.py [hunk 2]
```diff
@@ -74,8 +80,9 @@ def forward_gpu(self, inputs):
         out_h = conv.get_conv_outsize(h, kh, self.sy, self.ph)
         out_w = conv.get_conv_outsize(w, kw, self.sx, self.pw)
 
-        y = cuda.cupy.empty((n, out_c, out_h, out_w), dtype=x.dtype)
-        if cuda.cudnn_enabled and self.use_cudnn:
+        y = cuda.cupy.zeros((n, out_c, out_h, out_w), dtype=x.dtype)
```
Comment (delta2323) [inline chainer/functions/connection/convolution_2d.py]: Why do you initialize `y` with zero?

**Comment by delta2323** [comment]
Please merge or rebase the master branch after finishing the fix.

File: chainer/functions/connection/convolution_2d.py (hunk 1 above)
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: This is miss. I will remove it.

File: chainer/functions/connection/convolution_2d.py [hunk 3]
```diff
@@ -61,7 +66,8 @@ def forward_cpu(self, inputs):
         self.col = conv.im2col_cpu(
             x, kh, kw, self.sy, self.sx, self.ph, self.pw,
             cover_all=self.cover_all)
-        y = numpy.tensordot(self.col, W, ((1, 2, 3), (1, 2, 3)))
+        y = numpy.tensordot(
+            self.col, W, ((1, 2, 3), (1, 2, 3))).astype(x.dtype)
         if b is not None:
             y += b
```
Comment (delta2323) [inline chainer/functions/connection/convolution_2d.py]: Is it intentional that we do not convert `dtype` after we add `b` to `y`?

**Comment by delta2323** [comment]
ditto (don't we have to convert `dtype`?)
https://github.com/pfnet/chainer/pull/1162/files#diff-16a7c64c13ba043b2f7f21283bf70f47R162

File: chainer/utils/conv.py [hunk 4]
```diff
@@ -105,7 +105,7 @@ def col2im_gpu(col, sy, sx, ph, pw, h, w):
              for (int out_x = out_x_0; out_x < out_x_1; ++out_x) {
                int kx = x - out_x * sx;
                int k = out_y + out_h * (kx + kw * (ky + kh * c0));
-               val += col[out_x + out_w * k];
+               val = val + col[out_x + out_w * k];
```
Comment (delta2323) [inline chainer/utils/conv.py]: Could you tell me why the previous code does not work.

Comment (okuta) [inline chainer/utils/conv.py]: `+=` is not supported, if type of `val` is `float16`.

**Comment by delta2323** [comment]
I updated comments.

File: chainer/functions/connection/convolution_2d.py (hunk 3 above)
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: `b.dtype` and `y.dtype` are same.

**Comment by okuta** [comment]
Please merge #1212 before this PR.

File: chainer/functions/connection/convolution_2d.py (hunk 3 above)
Comment (delta2323) [inline chainer/functions/connection/convolution_2d.py]: OK, I understand it.

File: chainer/utils/conv.py (hunk 4 above)
Comment (delta2323) [inline chainer/utils/conv.py]: OK.

**Comment by delta2323** [comment]
LGTM 👍

**Comment by delta2323** [comment]
Now I had merged #1212. As changes of `cupy/core/carray.cuh` in this PR is different from #1212, please remove the change in this file from this PR.

**Comment by okuta** [comment]
I fixed.

**Comment by delta2323** [comment]
LGTM 👍

**Pull request merged**